### PR TITLE
Lie3D

### PR DESCRIPTION
**Pull request by endJunction**
Extension of the LIE implementation to 3D case.

TODO:
- [x] Testing the LIE/HM single fracture; comparison to the 2D case.

### CODE REVIEW

File: ProcessLib/LIE/HydroMechanics/HydroMechanicsProcessData.h [hunk 1]
```diff
@@ -128,10 +128,16 @@ struct HydroMechanicsProcessData
...
+    MeshLib::PropertyVector<double>* mesh_prop_stress_xz = nullptr;
     MeshLib::PropertyVector<double>* mesh_prop_strain_xx = nullptr;
     MeshLib::PropertyVector<double>* mesh_prop_strain_yy = nullptr;
+    MeshLib::PropertyVector<double>* mesh_prop_strain_zz = nullptr;
     MeshLib::PropertyVector<double>* mesh_prop_strain_xy = nullptr;
+    MeshLib::PropertyVector<double>* mesh_prop_strain_yz = nullptr;
+    MeshLib::PropertyVector<double>* mesh_prop_strain_xz = nullptr;
     MeshLib::PropertyVector<double>* mesh_prop_velocity = nullptr;
     MeshLib::PropertyVector<double>* mesh_prop_b = nullptr;
     MeshLib::PropertyVector<double>* mesh_prop_k_f = nullptr;
```
Comment (norihiro-w) [inline ProcessLib/LIE/HydroMechanics/HydroMechanicsProcessData.h]: please also add the followings
```
MeshLib::PropertyVector<double>* mesh_prop_w_s2 = nullptr;
MeshLib::PropertyVector<double>* mesh_prop_fracture_stress_shear2 = nullptr;
```

File: ProcessLib/LIE/HydroMechanics/LocalAssembler/HydroMechanicsLocalAssemblerFracture-impl.h [hunk 2]
```diff
@@ -369,8 +369,8 @@ computeSecondaryVariableConcreteWithVector(
 
     double ele_b = 0;
     double ele_k = 0;
-    Eigen::Vector2d ele_sigma_eff = Eigen::Vector2d::Zero();
-    Eigen::Vector2d ele_w = Eigen::Vector2d::Zero();
+    typename HMatricesType::ForceVectorType ele_sigma_eff = HMatricesType::ForceVectorType::Zero(GlobalDim);
+    typename HMatricesType::ForceVectorType ele_w = HMatricesType::ForceVectorType::Zero(GlobalDim);
     double ele_Fs = - std::numeric_limits<double>::max();
     for (auto const& ip : _ip_data)
     {
```
Comment (norihiro-w) [inline ProcessLib/LIE/HydroMechanics/LocalAssembler/HydroMechanicsLocalAssemblerFracture-impl.h]: please add the following at the very end of this function
```
 if (GlobalDim == 3)
 {
        (*_process_data.mesh_prop_w_s2)[element_id] = ele_w[1];
        (*_process_data.mesh_prop_fracture_stress_shear2)[element_id] = ele_sigma_eff[1];
 }
```

**Comment by norihiro-w** [comment]
please add mesh properties for fracture relative displacement and shear stress in  the additional coordinate. besides it looks good.

File: ProcessLib/LIE/SmallDeformation/LocalAssembler/SmallDeformationLocalAssemblerMatrix-impl.h [hunk 3]
```diff
@@ -172,31 +172,42 @@ void SmallDeformationLocalAssemblerMatrix<ShapeFunction, IntegrationMethod,
                                     DisplacementDim>::
 postTimestepConcrete(std::vector<double> const& /*local_x*/)
 {
-    const int n = 3;
-    std::valarray<double> ele_stress(0.0, n);
-    std::valarray<double> ele_strain(0.0, n);
+    // Compute average value per element
+    const int n = DisplacementDim == 2 ? 4 : 6;
+    Eigen::VectorXd ele_stress(n);
```
Comment (TomFischer) [inline ProcessLib/LIE/SmallDeformation/LocalAssembler/SmallDeformationLocalAssemblerMatrix-impl.h]: Why not `Eigen::VectorXd ele_stress = Eigen::VectorXd::Zero(n);`?

File: ProcessLib/LIE/SmallDeformation/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h [hunk 4]
```diff
@@ -270,31 +270,42 @@ void SmallDeformationLocalAssemblerMatrixNearFracture<ShapeFunction, Integration
                                     DisplacementDim>::
 postTimestepConcrete(std::vector<double> const& /*local_x*/)
 {
-    const int n = 3;
-    std::valarray<double> ele_stress(0.0, n);
-    std::valarray<double> ele_strain(0.0, n);
+    // Compute average value per element
+    const int n = DisplacementDim == 2 ? 4 : 6;
+    Eigen::VectorXd ele_stress(n);
```
Comment (TomFischer) [inline ProcessLib/LIE/SmallDeformation/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h]: Same question as above.

File: ProcessLib/LIE/SmallDeformation/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture.h [hunk 5]
```diff
@@ -158,6 +195,22 @@ class SmallDeformationLocalAssemblerMatrixNearFracture
         return cache;
     }
 
+    std::vector<double> const& getIntPtEpsilon(std::vector<double>& cache,
+                                             std::size_t const component) const
```
Comment (TomFischer) [inline ProcessLib/LIE/SmallDeformation/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture.h]: Indentation?

File: ProcessLib/LIE/SmallDeformation/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture.h [hunk 6]
```diff
@@ -158,6 +195,22 @@ class SmallDeformationLocalAssemblerMatrixNearFracture
...
+    {
+        cache.clear();
+        cache.reserve(_ip_data.size());
+
+        for (auto const& ip_data : _ip_data) {
```
Comment (TomFischer) [inline ProcessLib/LIE/SmallDeformation/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture.h]: Formatting.

File: ProcessLib/LIE/SmallDeformation/SmallDeformationProcess.cpp [hunk 7]
```diff
@@ -254,6 +342,25 @@ void SmallDeformationProcess<DisplacementDim>::initializeConcreteProcess(
     mesh_prop_epsilon_xy->resize(mesh.getNumberOfElements());
     _process_data._mesh_prop_strain_xy = mesh_prop_epsilon_xy;
 
+    if (DisplacementDim == 3)
+    {
+        auto mesh_prop_epsilon_xz =
```
Comment (TomFischer) [inline ProcessLib/LIE/SmallDeformation/SmallDeformationProcess.cpp]: Is it possible to use `MeshLib::getOrCreateMeshProperty`?

File: ProcessLib/LIE/SmallDeformation/LocalAssembler/SmallDeformationLocalAssemblerMatrix-impl.h (hunk 3 above)
Comment (endJunction) [inline ProcessLib/LIE/SmallDeformation/LocalAssembler/SmallDeformationLocalAssemblerMatrix-impl.h]: Thanks! This is indeed required.

**Comment by endJunction** [comment]
Thanks Tom! I've fixed the pointed issues. All of the changes are now also clang-formatted. Using only `getOrCreateMeshProperty` in the whole LIE implementation and postLIE.

@ all: For the suggested x,y,z tests and refinement there will be another PR.

**Comment by chleh** [comment]
> @ all: For the suggested x,y,z tests and refinement there will be another PR.

@endJunction, could you please open an issue for that, just for the record. Thanks.

**Pull request merged**